File: Week_04/127/word_ladder2.py
from typing import List


class Solution:
    def ladderLength(self, beginWord: str, endWord: str, wordList: List[str]) -> int:
        start_used = [False] * len(wordList)
        end_used = [False] * len(wordList)
        word_index_map = {w: i for i, w in enumerate(wordList)}
        if endWord not in word_index_map:
            return 0
        step = 1
        start_queue = [beginWord]
        end_queue = [endWord]
        end_used[word_index_map[endWord]] = True
        letters = [chr(o) for o in range(ord('a'), ord('z') + 1)]
        while start_queue:
            l = len(start_queue)
            step += 1
            for i in range(l):
                q = start_queue[i]
                q_char = list(q)
                for j in range(len(q)):
                    o = q_char[j]
                    for c in letters:
                        q_char[j] = c
                        s = ''.join(q_char)
                        if s in word_index_map:
                            if start_used[word_index_map[s]]:
                                continue
                            if end_used[word_index_map[s]]:
                                return step + 1
                            start_queue.append(s)
                            start_used[word_index_map[s]] = True
                    q_char[j] = o
            start_queue = start_queue[l:]
            if len(start_queue) > len(end_queue):
                start_queue, end_queue = end_queue, start_queue
                start_used, end_used = end_used, start_used
        return 0

File: Week_04/127/word_ladder2.go
package main

func ladderLength2(beginWord string, endWord string, wordList []string) int {
	startUsed := make([]bool, len(wordList))
	endUsed := make([]bool, len(wordList))

	wordMap := make(map[string]int, 0)
	for i, w := range wordList {
		wordMap[w] = i
	}
	if i, ok := wordMap[endWord]; !ok {
		return 0
	} else {
		endUsed[i] = true
	}
	startQueue := []string{beginWord}
	endQueue := []string{endWord}
	count := 1
	n := len(beginWord)
	for len(startQueue) > 0 {
		l := len(startQueue)
		count++
		for i := 0; i < l; i++ {
			chars := []byte(startQueue[i])
			for j := 0; j < n; j++ {
				o := chars[j]
				for c := 'a'; c <= 'z'; c++ {
					chars[j] = byte(c)
					idx, ok := wordMap[string(chars)]
					if !ok || startUsed[idx] {
						continue
					}
					if endUsed[idx] {
						return count
					}
					startUsed[idx] = true
					startQueue = append(startQueue, wordList[idx])
				}
				chars[j] = o
			}
		}
		startQueue = startQueue[l:]
		if len(startQueue) > len(endUsed) {
			startQueue, endQueue = endQueue, startQueue
			startUsed, endUsed = endUsed, startUsed
		}
	}
	return 0
}
